src/main/java/hudson/plugins/copyartifact/CopyArtifact.java

```diff
@@ -512,33 +523,112 @@ private static ItemGroup getItemGroup(Run<?, ?> build) {
     }
 
 
-    private boolean perform(Run src, Run<?,?> dst, String expandedFilter, String expandedExcludes, FilePath targetDir,
-            FilePath baseTargetDir, Copier copier, PrintStream console)
-            throws IOException, InterruptedException {
-        FilePath srcDir = selector.getSourceDirectory(src, console);
+    private boolean perform(Run src, Run<?,?> dst, String expandedFilter, @CheckForNull String expandedExcludes, FilePath targetDir, PrintStream console) throws IOException, InterruptedException {
+        VirtualFile srcDir = selector.getArtifacts(src, console);
         if (srcDir == null) {
             return isOptional();  // Fail build unless copy is optional
         }
 
-        copier.initialize(src, dst, srcDir, baseTargetDir);
+        Map<String, String> fingerprints;
+        MessageDigest md5;
+        if (isFingerprintArtifacts()) {
+            fingerprints = new HashMap<>();
+            try {
+                md5 = MessageDigest.getInstance("MD5");
+            } catch (NoSuchAlgorithmException x) {
+                throw new AssertionError(x);
```

Comment:
does not really matter, but maybe just IOException? There were some discussions about disabling MD5 at all to harden the security. I'd guess it is not realistic, but I would rather prefer to avoid errors in the production code

Reply:
To do that we would first need to reimplement fingerprints in Jenkins core, which would surely force changes here as well anyway.